titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/ProtobufCache.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2021 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.titus.runtime.endpoint.v3.grpc;
+
+import java.time.Duration;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.netflix.spectator.api.Counter;
+import com.netflix.spectator.api.Registry;
+import com.netflix.spectator.api.patterns.PolledMeter;
+import com.netflix.titus.common.framework.scheduler.ScheduleReference;
+import com.netflix.titus.common.framework.scheduler.model.ScheduleDescriptor;
+import com.netflix.titus.common.runtime.TitusRuntime;
+import com.netflix.titus.common.util.CollectionsExt;
+import com.netflix.titus.common.util.time.Clock;
+
+class ProtobufCache<E, C> {
+
+    private static final String METRICS_ROOT = "protobufCache.";
+
+    private final Function<E, C> converter;
+    private final Supplier<Predicate<E>> expiredPredicateSupplier;
+    private final GrpcObjectsCacheConfiguration configuration;
+    private final Clock clock;
+
+    @VisibleForTesting
+    final ConcurrentMap<String, CacheEntry<E, C>> cache = new ConcurrentHashMap<>();
```

Comment:
did you consider using a Guava cache with capacity-based LRU eviction, instead of relying only on time-based periodic evictions? It would be more resilient to memory blowing up and potentially simpler implementation?

Reply:
I wanted to use Caffeine, but in the end I decided that it will be better to write a custom cache implementation.